app/src/main/java/edu/wisc/ece/pinpoint/utils/FirebaseDriver.java

```diff
@@ -350,6 +394,83 @@ public Task<Void> deletePin(String pid) {
         });
     }
 
+    public HashSet<String> getCachedFollowerIds() {return followerIds;}
+
+    public HashSet<String> getCachedFollowingIds() {return followingIds;}
+
+    public void followUser(String uid) {
+        if (auth.getUid() == null) {
+            throw new IllegalStateException("User must be logged in to follow an account");
+        }
+        Date timestamp = new Date();
+        followingIds.add(uid);
+        // Add target to following
+        db.collection("social")
+                .document(auth.getUid()).update("following", FieldValue.arrayUnion(uid))
+                .addOnSuccessListener(complete -> db.collection("users").document(
+                        auth.getUid()).update("numFollowing", FieldValue.increment(1)));
+
+
+        // Add self to target's followers
+        db.collection("social")
+                .document(uid).update("followers", FieldValue.arrayUnion(auth.getUid()))
+                .addOnSuccessListener(complete -> db.collection("users").document(uid)
+                        .update("numFollowers", FieldValue.increment(1)));
```

Comment:
Can we do this as a batched write to ensure it all gets updated or none of it gets updated? And then also have the local cache update only execute after the batched write completes successfully. On failure we can show a Toast and log the error. Here's documentation on how to do a batched write in Android: https://firebase.google.com/docs/firestore/manage-data/transactions#batched-writes